rust/kernel/xarray.rs

```diff
@@ -0,0 +1,290 @@
+        let old = unsafe {
+            bindings::xa_store(
+                self.xa.get(),
+                index.try_into()?,
+                new as *mut _,
+                bindings::GFP_KERNEL,
+            )
+        };
+
+        let ret = unsafe { bindings::xa_err(old) };
+        if ret != 0 {
+            Err(Error::from_errno(ret))
+        } else if old.is_null() {
+            guard.dismiss();
+            Ok(None)
+        } else {
+            guard.dismiss();
+            // SAFETY: The old value must have been stored by either this function or
+            // `alloc_limits_opt`, both of which ensure non-NULL entries are valid
+            // ForeignOwnable pointers.
+            Ok(Some(unsafe { T::from_foreign(old) }))
+        }
+    }
+
+    /// Replaces an entry with a new value, dropping the old value (if any).
+    pub fn set(self: Pin<&Self>, index: usize, value: T) -> Result {
+        self.replace(index, value)?;
+        Ok(())
+    }
+
+    /// Looks up a reference to an entry in the array, cloning it
+    /// and returning the cloned value to the user.
+    ///
+    pub fn get(self: Pin<&Self>, index: u64) -> Option<T::Borrowed<'_>>
+    where
+        T: Clone,
+    {
+        // SAFETY: `self.xa` is always valid by the type invariant.
+        unsafe { bindings::xa_lock(self.xa.get()) };
+
+        // SAFETY: `self.xa` is always valid by the type invariant.
+        let p = unsafe { bindings::xa_load(self.xa.get(), index) } as *const T;
+
+        let t = NonNull::new(p as *mut T).map(|p| unsafe { T::borrow(p.clone().as_ptr() as _) });
+
+        // SAFETY: `self.xa` is always valid by the type invariant.
+        unsafe { bindings::xa_unlock(self.xa.get()) };
+
+        t
+    }
+
+    /// Looks up and a reference to an entry in the array, calling the user
+    /// provided function on the resulting `Option<&T>` to return a value
+    /// computed from the reference. Use this function if you need shared
+    /// access to a `&T` that is not `Clone`.
+    ///
+    pub fn get_shared<F, R>(self: Pin<&Self>, index: u64, f: F) -> R
+    where
+        F: FnOnce(Option<T::Borrowed<'_>>) -> R,
+    {
+        // SAFETY: `self.xa` is always valid by the type invariant.
+        unsafe { bindings::xa_lock(self.xa.get()) };
+
+        // SAFETY: `self.xa` is always valid by the type invariant.
+        let p = unsafe { bindings::xa_load(self.xa.get(), index) };
+        let t = NonNull::new(p as *mut T).map(|p| unsafe { T::borrow(p.as_ptr() as _) });
+        let r = f(t);
+
+        // SAFETY: `self.xa` is always valid by the type invariant.
+        unsafe { bindings::xa_unlock(self.xa.get()) };
+
+        r
+    }
+
+    /// Looks up and returns a *mutable* reference to an entry in the array.
+    pub fn get_mutable(self: Pin<&mut Self>, index: u64) -> Option<ScopeGuard<T, fn(T)>> {
+        // SAFETY: `self.xa` is always valid by the type invariant.
+        unsafe { bindings::xa_lock(self.xa.get()) };
```

Comment:
This probably depends on changes to `borrow_mut`. I think ideally we want the trait to have a `BorrowedMut` type that represents the kind of access you can get to the inner value given mutable access to the `T`, then return `Option<T::BorrowedMut<'a>>`.

For an `Arc`, the `BorrowedMut` type would still be `ArcBorrow`, but that actually still gives this extra powers over the two other methods. It lets you access the `Arc`s contents without modifying the refcount, and without a closure.

And of course, with a `Box`, the `BorrowedMut` type would be `&mut T`, so this type would return `Option<&mut T>` when the pointer type is `Box`.